cli/command/system/inspect.go

```diff
@@ -103,38 +102,53 @@ func inspectPlugin(ctx context.Context, dockerCli *command.DockerCli) inspect.Ge
 
 func inspectAll(ctx context.Context, dockerCli *command.DockerCli, getSize bool, typeConstraint string) inspect.GetRefFunc {
 	var inspectAutodetect = []struct {
-		ObjectType      string
-		IsSizeSupported bool
-		ObjectInspector func(string) (interface{}, []byte, error)
+		objectType      string
+		isSizeSupported bool
+		isSwarmObject   bool
+		objectInspector func(string) (interface{}, []byte, error)
 	}{
-		{"container", true, inspectContainers(ctx, dockerCli, getSize)},
-		{"image", false, inspectImages(ctx, dockerCli)},
-		{"network", false, inspectNetwork(ctx, dockerCli)},
-		{"volume", false, inspectVolume(ctx, dockerCli)},
-		{"service", false, inspectService(ctx, dockerCli)},
-		{"task", false, inspectTasks(ctx, dockerCli)},
-		{"node", false, inspectNode(ctx, dockerCli)},
-		{"plugin", false, inspectPlugin(ctx, dockerCli)},
+		{"container", true, false, inspectContainers(ctx, dockerCli, getSize)},
+		{"image", false, false, inspectImages(ctx, dockerCli)},
+		{"network", false, false, inspectNetwork(ctx, dockerCli)},
+		{"volume", false, false, inspectVolume(ctx, dockerCli)},
+		{"service", false, true, inspectService(ctx, dockerCli)},
+		{"task", false, true, inspectTasks(ctx, dockerCli)},
+		{"node", false, true, inspectNode(ctx, dockerCli)},
+		{"plugin", false, false, inspectPlugin(ctx, dockerCli)},
 	}
 
-	isErrNotSwarmManager := func(err error) bool {
-		return strings.Contains(err.Error(), "This node is not a swarm manager")
+	// isSwarmManager does an Info API call to verify that the daemon is
+	// a swarm manager.
+	isSwarmManager := func() bool {
+		info, err := dockerCli.Client().Info(ctx)
+		if err != nil {
+			fmt.Fprintf(dockerCli.Err(), "%v\n", err)
+			return false
+		}
+		return info.Swarm.ControlAvailable
 	}
 
 	return func(ref string) (interface{}, []byte, error) {
+		// assume node is an active swarm manager until proven otherwise
+		isSwarmSupported := true
+
 		for _, inspectData := range inspectAutodetect {
-			if typeConstraint != "" && inspectData.ObjectType != typeConstraint {
+			if typeConstraint != "" && inspectData.objectType != typeConstraint {
+				continue
+			}
+			if typeConstraint == "" && inspectData.isSwarmObject && (!isSwarmSupported || !isSwarmManager()) {
```

Comment:
Let's make sure we don't call `isSwarmManager` multiple times.

Perhaps something like the following:

```go
const (
        swarmSupportUnknown = iota
        swarmSupported
        swarmUnsupported
)

isSwarmSupported := swarmSupportUnknown

...

if typeConstraint == "" && inspectData.isSwarmObject {
        if isSwarmSupported == swarmUnsupported || (isSwarmSupported == swarmSupportUnknown && !isSwarmManager()) {
                isSwarmSupported = swarmUnsupported
                continue
        }
}
```